semgrep/semgrep/target_manager.py

```diff
@@ -0,0 +1,234 @@
+
+        self._filtered_targets: Dict[str, Set[Path]] = {}
+
+    @staticmethod
+    def resolve_targets(targets: List[str]) -> Set[Path]:
+        """
+            Return list of Path objects appropriately resolving relative paths
+            (relative to cwd) if necessary
+        """
+        base_path = Path(".")
+        return set(
+            Path(target) if Path(target).is_absolute() else base_path.joinpath(target)
+            for target in targets
+        )
+
+    @staticmethod
+    def _parse_output(output: str, curr_dir: Path) -> Set[Path]:
+        """
+            Convert a newline delimited list of files to a set of path objects
+            prepends curr_dir to all paths in said list
+
+            If list is empty then returns an empty set
+        """
+        files: Set[Path] = set()
+        if output:
+            files = set(Path(curr_dir) / elem for elem in output.strip().split("\n"))
+        return files
+
+    @staticmethod
+    def _expand_dir(
+        curr_dir: Path, language: str, visible_to_git_only: bool
+    ) -> Set[Path]:
+        """
+            Recursively go through a directory and return list of all files with
+            default file extention of language
+        """
+        extensions = lang_to_exts(language)
+        expanded: Set[Path] = set()
+
+        for ext in extensions:
+            if visible_to_git_only:
+                try:
+                    # Tracked files
+                    tracked_output = subprocess.check_output(
+                        ["git", "ls-files", f"*.{ext}"],
+                        cwd=curr_dir.resolve(),
+                        encoding="utf-8",
+                        stderr=subprocess.DEVNULL,
+                    )
+
+                    # Untracked but not ignored files
+                    untracked_output = subprocess.check_output(
+                        [
+                            "git",
+                            "ls-files",
+                            "--other",
+                            "--exclude-standard",
+                            f"*.{ext}",
+                        ],
+                        cwd=curr_dir.resolve(),
+                        encoding="utf-8",
+                        stderr=subprocess.DEVNULL,
+                    )
+                except subprocess.CalledProcessError:
+                    raise NotGitProjectError(
+                        f"{curr_dir.resolve()} is not a git repository."
+                    )
+
+                tracked = TargetManager._parse_output(tracked_output, curr_dir)
+                untracked_unignored = TargetManager._parse_output(
+                    untracked_output, curr_dir
+                )
+
+                expanded = expanded.union(tracked)
+                expanded = expanded.union(untracked_unignored)
+
+            else:
+                output = subprocess.run(
+                    ["find", curr_dir, "-type", "f", "-name", f"*.{ext}"],
```

Comment:
shouldn't we be using `Path(.).rglob("*")`?

Reply:
This is exactly what semgrep-core does right now

Reply:
```
python3 -c "from pathlib import Path; [p for p in Path('..').rglob('*.py')]"  0.06s user 0.06s system 96% cpu 0.129 total
python3 -c "from pathlib import Path;"  0.03s user 0.01s system 90% cpu 0.044 total
```

Quick and dirty experiment is using glob takes 10x more time (70 ms vs 7ms) for the small number of files in semgrep

Reply:
amazing

Reply:
`time python3 -c 'import subprocess; subprocess.run(["find", "..", "-type", "f", "-name", f"*.py"])'  0.05s user 0.08s system 69% cpu 0.112 total`
With subprocess overhead they're around equivalent.